### DISCUSSION TITLE
UPDITERM: Serial debugging directly via serialUPDI (no extra HW required, not using ATtiny Rx/Tx)

### DISCUSSION THREAD

**Discussion by ArnieO** [SpenceKonde/megaTinyCore#1163] [Show and tell]
I just stumbled over this nice tool that enables me to do printf() from Arduino code to do debugging:
https://github.com/streef/upditerm

The author writes:
_Upditerm is a terminal program similar to miniterm, but instead of connecting directly to a serial port, it uses a serial port wired as a UPDI interface to access a virtual UART on the target CPU. You can use upditerm to interact with a program on an AVR microcontroller using only the UPDI interface._

As it was not evident for me how to use it, I wrote this: https://github.com/streef/upditerm/issues/1#issue-2635916100

**Comment by hmeijdam** [comment]
I found it a nice trick. We discussed it in [this discussion](https://github.com/SpenceKonde/megaTinyCore/discussions/342)

**Reply by ArnieO** [reply]
Thanks! I did a search yesterday before posting but did not find that discussion.
